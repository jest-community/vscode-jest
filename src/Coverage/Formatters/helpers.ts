import { Location, Range } from 'istanbul-lib-coverage';

export function isValidPosition(p: Location): boolean {
  return (p || false) && p.line !== null && p.line >= 0;
}

export function isValidLocation(l: Range): boolean {
  return isValidPosition(l.start) && isValidPosition(l.end);
}
