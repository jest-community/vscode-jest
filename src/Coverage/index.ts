export * from './CoverageMapProvider';
export * from './CoverageOverlay';
export * from './CoverageCodeLensProvider';
