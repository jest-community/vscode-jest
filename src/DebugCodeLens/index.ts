export { DebugCodeLensProvider } from './DebugCodeLensProvider'
export { TestState } from './TestState'
