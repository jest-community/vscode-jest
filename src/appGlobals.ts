export const extensionName = 'io.orta.jest'
export const extensionId = 'orta.vscode-jest'
