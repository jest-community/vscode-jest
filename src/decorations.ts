import { window, OverviewRulerLane, DecorationRangeBehavior } from 'vscode'

export function failingItName() {
  return window.createTextEditorDecorationType({
    overviewRulerColor: 'red',
    overviewRulerLane: OverviewRulerLane.Left,
    light: {
      before: {
        color: '#FF564B',
        contentText: '✘ ',
      },
    },
    dark: {
      before: {
        color: '#AD322D',
        contentText: '✘ ',
      },
    },
    rangeBehavior: DecorationRangeBehavior.ClosedClosed,
  })
}

export function skipItName() {
  return window.createTextEditorDecorationType({
    overviewRulerColor: 'yellow',
    overviewRulerLane: OverviewRulerLane.Left,
    light: {
      before: {
        color: '#fed37f',
        contentText: '○ ',
      },
    },
    dark: {
      before: {
        color: '#fed37f',
        contentText: '○ ',
      },
    },
    rangeBehavior: DecorationRangeBehavior.ClosedClosed,
  })
}

export function passingItName() {
  return window.createTextEditorDecorationType({
    overviewRulerColor: 'green',
    overviewRulerLane: OverviewRulerLane.Left,
    light: {
      before: {
        color: '#3BB26B',
        contentText: '✔ ',
      },
    },
    dark: {
      before: {
        color: '#2F8F51',
        contentText: '✔ ',
      },
    },
    rangeBehavior: DecorationRangeBehavior.ClosedClosed,
  })
}

export function notRanItName() {
  return window.createTextEditorDecorationType({
    overviewRulerColor: 'darkgrey',
    overviewRulerLane: OverviewRulerLane.Left,
    dark: {
      before: {
        color: '#9C9C9C',
        contentText: '○ ',
      },
    },
    light: {
      before: {
        color: '#7C7C7C',
        contentText: '○ ',
      },
    },
    rangeBehavior: DecorationRangeBehavior.ClosedClosed,
  })
}

export function failingAssertionStyle(text: string) {
  return window.createTextEditorDecorationType({
    isWholeLine: true,
    overviewRulerColor: 'red',
    overviewRulerLane: OverviewRulerLane.Left,
    light: {
      before: {
        color: '#FF564B',
      },
      after: {
        color: '#FF564B',
        contentText: ' // ' + text,
      },
    },
    dark: {
      before: {
        color: '#AD322D',
      },
      after: {
        color: '#AD322D',
        contentText: ' // ' + text,
      },
    },
  })
}
