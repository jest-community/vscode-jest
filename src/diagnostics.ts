/**
 * this module contains functions to show jest test results in
 * vscode inspector via the DiagnosticsCollection.
 */
import * as vscode from 'vscode'
import { existsSync } from 'fs'
// import { DiagnosticCollection, Uri, Diagnostic, Range, DiagnosticSeverity } from 'vscode'
import { TestFileAssertionStatus } from 'jest-editor-support'
import { TestReconciliationState } from './TestResults'

export function updateDiagnostics(testResults: TestFileAssertionStatus[], diagnostics: vscode.DiagnosticCollection) {
  function addTestFileError(result: TestFileAssertionStatus, uri: vscode.Uri) {
    const diag = new vscode.Diagnostic(
      new vscode.Range(0, 0, 0, 0),
      result.message || 'test file error',
      vscode.DiagnosticSeverity.Error
    )
    diag.source = 'Jest'
    diagnostics.set(uri, [diag])
  }
  function addTestsError(result: TestFileAssertionStatus, uri: vscode.Uri) {
    const asserts = result.assertions.filter(a => a.status === TestReconciliationState.KnownFail)
    diagnostics.set(
      uri,
      asserts.map(assertion => {
        let line: number
        if (assertion.line >= 0) {
          line = Math.max(assertion.line - 1, 0)
        } else {
          line = 0
          console.warn(
            `received invalid line number '${assertion.line}' for '${uri.toString()}'. (most likely due to unexpected test results... you can help fix the root cause by logging an issue with a sample project to reproduce this warning)`
          )
        }
        let startColumn = 0
        let endColumn = Number.MAX_SAFE_INTEGER
        const associatedTextEditor = vscode.window.visibleTextEditors.find(e => e.document.uri.fsPath === uri.fsPath)
        if (associatedTextEditor) {
          const textLine = associatedTextEditor.document.lineAt(line)
          startColumn = textLine.firstNonWhitespaceCharacterIndex
          endColumn = textLine.text.length
        }
        const diag = new vscode.Diagnostic(
          new vscode.Range(line, startColumn, line, endColumn),
          assertion.terseMessage || assertion.shortMessage || assertion.message,
          vscode.DiagnosticSeverity.Error
        )
        diag.source = 'Jest'
        return diag
      })
    )
  }

  testResults.forEach(result => {
    const uri = vscode.Uri.file(result.file)
    switch (result.status) {
      case TestReconciliationState.KnownFail:
        if (result.assertions.length <= 0) {
          addTestFileError(result, uri)
        } else {
          addTestsError(result, uri)
        }
        break
      default:
        diagnostics.delete(uri)
        break
    }
  })

  // Remove diagnostics for files no longer in existence
  const toBeDeleted = []
  diagnostics.forEach(uri => {
    if (!existsSync(uri.fsPath)) {
      toBeDeleted.push(uri)
    }
  })
  toBeDeleted.forEach(uri => {
    diagnostics.delete(uri)
  })
}

export function resetDiagnostics(diagnostics: vscode.DiagnosticCollection) {
  diagnostics.clear()
}
export function failedSuiteCount(diagnostics: vscode.DiagnosticCollection): number {
  let sum = 0
  diagnostics.forEach(() => sum++)
  return sum
}
