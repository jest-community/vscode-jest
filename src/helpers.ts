import { platform } from 'os';
import { existsSync, readFileSync, mkdirSync, writeFileSync } from 'fs';
import { normalize, join } from 'path';
import { ExtensionContext } from 'vscode';

import { PluginResourceSettings, hasUserSetPathToJest } from './Settings';
import { TestIdentifier } from './TestResults';
import { TestStats } from './types';

/**
 * Known binary names of `react-scripts` forks
 */
const createReactAppBinaryNames = [
  'react-scripts',
  'react-native-scripts',
  'react-scripts-ts',
  'react-app-rewired',
];

/**
 * File extension for npm binaries
 */
export const nodeBinExtension: string = platform() === 'win32' ? '.cmd' : '';

/**
 * Resolves the location of an npm binary
 *
 * Returns the path if it exists, or `undefined` otherwise
 */
function getLocalPathForExecutable(rootPath: string, executable: string): string | undefined {
  const absolutePath = normalize(
    join(rootPath, 'node_modules', '.bin', executable + nodeBinExtension)
  );
  return existsSync(absolutePath) ? absolutePath : undefined;
}

/**
 * Tries to read the test command from the scripts section within `package.json`
 *
 * Returns the test command in case of success,
 * `undefined` if there was an exception while reading and parsing `package.json`
 * `null` if there is no test script
 */
export function getTestCommand(rootPath: string): string | undefined | null {
  try {
    const packagePath = join(rootPath, 'package.json');
    const packageJSON = JSON.parse(readFileSync(packagePath, 'utf8'));
    if (packageJSON && packageJSON.scripts && packageJSON.scripts.test) {
      return packageJSON.scripts.test;
    }
    return null;
  } catch {
    return undefined;
  }
}

/**
 * Checks if the supplied test command could have been generated by create-react-app
 */
export function isCreateReactAppTestCommand(testCommand?: string | null): boolean {
  return (
    !!testCommand &&
    createReactAppBinaryNames.some((binary) => testCommand.includes(`${binary} test`))
  );
}

/**
 * Checks if the project in `rootPath` was bootstrapped by `create-react-app`.
 */
function isBootstrappedWithCreateReactApp(rootPath: string): boolean {
  const testCommand = getTestCommand(rootPath);
  if (testCommand === undefined) {
    // In case parsing `package.json` failed or was unconclusive,
    // fallback to checking for the presence of the binaries in `./node_modules/.bin`
    return createReactAppBinaryNames.some(
      (binary) => getLocalPathForExecutable(rootPath, binary) !== undefined
    );
  }
  return isCreateReactAppTestCommand(testCommand);
}

/**
 * Handles getting the jest runner, handling the OS and project specific work too
 *
 * @returns {string}
 */
// tslint:disable-next-line no-shadowed-variable
export function pathToJest({ pathToJest, rootPath }: PluginResourceSettings): string {
  if (pathToJest && hasUserSetPathToJest(pathToJest)) {
    return normalize(pathToJest);
  }

  if (isBootstrappedWithCreateReactApp(rootPath)) {
    return 'npm test --';
  }

  const p = getLocalPathForExecutable(rootPath, 'jest') || 'jest' + nodeBinExtension;
  return `"${p}"`;
}

/**
 * Handles getting the path to config file
 *
 * @returns {string}
 */
export function pathToConfig(pluginSettings: PluginResourceSettings): string {
  if (pluginSettings.pathToConfig) {
    return normalize(pluginSettings.pathToConfig);
  }

  return '';
}

/**
 *  Taken From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
 */
export function escapeRegExp(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
}

/**
 * ANSI colors/characters cleaning based on http://stackoverflow.com/questions/25245716/remove-all-ansi-colors-styles-from-strings
 */
export function cleanAnsi(str: string): string {
  return str.replace(
    // eslint-disable-next-line no-control-regex
    /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g,
    ''
  );
}

export type IdStringType = 'display' | 'display-reverse' | 'full-name';
export function testIdString(type: IdStringType, identifier: TestIdentifier): string {
  if (!identifier.ancestorTitles.length) {
    return identifier.title;
  }
  const parts = [...identifier.ancestorTitles, identifier.title];
  switch (type) {
    case 'display':
      return parts.join(' > ');
    case 'display-reverse':
      return parts.reverse().join(' < ');
    case 'full-name':
      return parts.join(' ');
  }
}

/** convert the upper-case drive letter filePath to lower-case. If path does not contain upper-case drive letter, returns undefined. */
// note: this should probably be replaced by vscode.URI.file(filePath).fsPath ...
export function toLowerCaseDriveLetter(filePath: string): string | undefined {
  const match = filePath.match(/^([A-Z]:\\)(.*)$/);
  if (match) {
    return `${match[1].toLowerCase()}${match[2]}`;
  }
}
/** convert the lower-case drive letter filePath (like vscode.URI.fsPath) to lower-case. If path does not contain lower-case drive letter, returns undefined. */
export function toUpperCaseDriveLetter(filePath: string): string | undefined {
  const match = filePath.match(/^([a-z]:\\)(.*)$/);
  if (match) {
    return `${match[1].toUpperCase()}${match[2]}`;
  }
}

/**
 * convert vscode.URI.fsPath to the actual file system file-path, i.e. convert drive letter to upper-case for windows
 * @param filePath
 */
export function toFilePath(filePath: string): string {
  return toUpperCaseDriveLetter(filePath) || filePath;
}

/**
 * Generate path to icon used in decorations
 * NOTE: Should not be called repeatedly for the performance reasons. Cache your results.
 */
export function prepareIconFile(
  context: ExtensionContext,
  iconName: string,
  source: string,
  color?: string
): string {
  const iconsPath = join('generated-icons');

  const resolvePath = (...args: string[]): string => {
    return context.asAbsolutePath(join(...args));
  };

  const resultIconPath = resolvePath(iconsPath, `${iconName}.svg`);
  let result = source.toString();

  if (color) {
    result = result.replace('fill="currentColor"', `fill="${color}"`);
  }

  if (!existsSync(resultIconPath) || readFileSync(resultIconPath).toString() !== result) {
    if (!existsSync(resolvePath(iconsPath))) {
      mkdirSync(resolvePath(iconsPath));
    }

    writeFileSync(resultIconPath, result);
  }

  return resultIconPath;
}

const SurroundingQuoteRegex = /^["']+|["']+$/g;
export const removeSurroundingQuote = (command: string): string =>
  command.replace(SurroundingQuoteRegex, '');

// TestStats
export const emptyTestStats = (): TestStats => {
  return { success: 0, fail: 0, unknown: 0 };
};
