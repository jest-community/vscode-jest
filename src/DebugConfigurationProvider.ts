import * as vscode from 'vscode'
import { getTestCommand, isCRATestCommand } from './helpers'

export class DebugConfigurationProvider implements vscode.DebugConfigurationProvider {
  private fileNameToRun: string = ''
  private testToRun: string = ''

  /**
   * Prepares injecting the name of the test, which has to be debugged, into the `DebugConfiguration`,
   * This function has to be called before `vscode.debug.startDebugging`.
   */
  public prepareTestRun(fileNameToRun: string, testToRun: string) {
    this.fileNameToRun = fileNameToRun
    this.testToRun = testToRun
  }

  resolveDebugConfiguration(
    _folder: vscode.WorkspaceFolder | undefined,
    debugConfiguration: vscode.DebugConfiguration,
    _token?: vscode.CancellationToken
  ) {
    if (debugConfiguration.name !== 'vscode-jest-tests') {
      return debugConfiguration
    }

    if (!debugConfiguration.env) {
      debugConfiguration.env = {}
    }
    // necessary for running CRA test scripts in non-watch mode
    debugConfiguration.env.CI = 'vscode-jest-tests'

    if (this.fileNameToRun) {
      debugConfiguration.args.push(this.fileNameToRun)
      if (this.testToRun) {
        debugConfiguration.args.push('--testNamePattern')
        debugConfiguration.args.push(this.testToRun)
      }
      this.fileNameToRun = ''
      this.testToRun = ''
    }

    return debugConfiguration
  }

  provideDebugConfigurations(folder: vscode.WorkspaceFolder | undefined, _token?: vscode.CancellationToken) {
    // default jest config according to:
    // https://github.com/Microsoft/vscode-recipes/tree/master/debugging-jest-tests#configure-launchjson-file-for-your-test-framework

    // create-react-app config according to:
    // https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#debugging-tests-in-visual-studio-code

    const debugConfiguration: vscode.DebugConfiguration = {
      type: 'node',
      name: 'vscode-jest-tests',
      request: 'launch',
      args: ['--runInBand'],
      cwd: '${workspaceFolder}',
      console: 'integratedTerminal',
      internalConsoleOptions: 'neverOpen',
    }

    const testCommand = folder && getTestCommand(folder.uri.fsPath)
    if (isCRATestCommand(testCommand)) {
      const craCommand = testCommand.split(' ')
      // Settings specific for projects bootstrapped with `create-react-app`
      debugConfiguration.runtimeExecutable = '${workspaceFolder}/node_modules/.bin/' + craCommand.shift()
      debugConfiguration.args = [...craCommand, ...debugConfiguration.args]
      debugConfiguration.protocol = 'inspector'
    } else if (testCommand === 'node scripts/test.js --env=jsdom') {
      // Settings specific for projects bootstrapped with `create-react-app` after being ejected
      const craCommand = testCommand.split(' ')
      craCommand.shift()
      debugConfiguration.program = '${workspaceFolder}/' + craCommand.shift()
      debugConfiguration.args = [...craCommand, ...debugConfiguration.args]
      debugConfiguration.protocol = 'inspector'
    } else {
      // Plain jest setup
      debugConfiguration.program = '${workspaceFolder}/node_modules/jest/bin/jest'
    }

    return [debugConfiguration]
  }
}
