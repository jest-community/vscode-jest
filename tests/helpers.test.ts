jest.unmock('../src/helpers')
jest.unmock('../src/Settings')

const mockExistsSync = jest.fn()
const mockReadFileSync = jest.fn()
jest.mock('fs', () => ({
  existsSync: mockExistsSync,
  readFileSync: mockReadFileSync,
}))

const mockPlatform = jest.fn()
jest.mock('os', () => ({ platform: mockPlatform }))

const mockJoin = jest.fn()
const mockNormalize = jest.fn()
jest.mock('path', () => ({
  join: mockJoin,
  normalize: mockNormalize,
}))

import { isCreateReactAppTestCommand, pathToJest, pathToJestPackageJSON } from '../src/helpers'
import * as path from 'path'

describe('ModuleHelpers', () => {
  describe('pathToJestPackageJSON', () => {
    const defaultPathToJest = null
    const defaultSettings: any = {
      pathToJest: defaultPathToJest,
      rootPath: '',
    }

    beforeEach(() => {
      mockJoin.mockImplementation(require.requireActual('path').join)
      mockNormalize.mockImplementation(require.requireActual('path').normalize)
    })

    it('should return null when not found', () => {
      mockExistsSync.mockReturnValue(false)

      expect(pathToJestPackageJSON(defaultSettings)).toBe(null)
    })

    describe('pathToJest: <default>', () => {
      // tslint:disable no-shadowed-variable
      describe('rootPath: <default>', () => {
        it('should return package.json when Jest is installed as a dependency', () => {
          const expected = path.join('node_modules', 'jest', 'package.json')
          mockExistsSync.mockImplementation(path => path === expected)

          expect(pathToJestPackageJSON(defaultSettings)).toBe(expected)
        })

        it('should return package.json when React Scripts are installed as a dependency', () => {
          const expected = path.join('node_modules', 'react-scripts', 'node_modules', 'jest', 'package.json')
          mockExistsSync.mockImplementation(path => path === expected)

          expect(pathToJestPackageJSON(defaultSettings)).toBe(expected)
        })
      })

      describe('rootPath: <string>', () => {
        it('should return package.json when Jest is installed as a dependency', () => {
          const expected = path.join('..', '..', 'node_modules', 'jest', 'package.json')
          mockExistsSync.mockImplementation(path => path === expected)

          const workspace: any = {
            pathToJest: defaultPathToJest,
            rootPath: path.join('..', '..'),
          }
          expect(pathToJestPackageJSON(workspace)).toBe(expected)
        })

        it('should return package.json when Jest-cli is installed as a dependency', () => {
          const expected = path.join('..', '..', 'node_modules', 'jest-cli', 'package.json')
          mockExistsSync.mockImplementation(path => path === expected)

          const workspace: any = {
            rootPath: path.join('..', '..'),
            pathToJest: defaultPathToJest,
          }
          expect(pathToJestPackageJSON(workspace)).toBe(expected)
        })

        it('should return package.json when React Scripts are installed as a dependency', () => {
          const expected = path.join(
            '..',
            '..',
            'node_modules',
            'react-scripts',
            'node_modules',
            'jest',
            'package.json'
          )
          mockExistsSync.mockImplementation(path => path === expected)

          const workspace: any = {
            rootPath: path.join('..', '..'),
            pathToJest: '',
          }
          expect(pathToJestPackageJSON(workspace)).toBe(expected)
        })
      })
    })

    describe('pathToJest: npm test --', () => {
      it('will not find package.json', () => {
        const expected = null
        mockExistsSync.mockImplementation(path => path === expected)

        const workspace: any = {
          rootPath: '',
          pathToJest: 'npm test --',
        }
        expect(pathToJestPackageJSON(workspace)).toBe(null)
      })
    })

    describe(`pathToJest: ${path.join('"test scripts"', 'test')}`, () => {
      it('will not find package.json', () => {
        const expected = path.join('"test scripts"', 'test')
        mockExistsSync.mockImplementation(path => path === expected)

        const workspace: any = {
          rootPath: '',
          pathToJest: expected,
        }
        expect(pathToJestPackageJSON(workspace)).toBe(null)
      })
    })
  })

  describe('isCreateReactAppTestCommand', () => {
    it('should return true for CRA', () => {
      expect(isCreateReactAppTestCommand('react-scripts test --env=jsdom')).toBe(true)
    })

    it('should return false for other scripts', () => {
      expect(isCreateReactAppTestCommand('custom-script')).toBe(false)
    })
  })

  describe('pathToJest', () => {
    const defaultSettings: any = {
      pathToJest: null,
      rootPath: '',
    }

    it('returns "npm test --" when bootstrapped with create-react-app', () => {
      mockReadFileSync.mockReturnValueOnce(
        JSON.stringify({
          scripts: {
            test: 'react-scripts test',
          },
        })
      )

      expect(pathToJest(defaultSettings)).toBe('npm test --')
    })

    it('returns the normalized "pathToJest" setting when set by the user', () => {
      const expected = {}
      mockNormalize.mockReturnValueOnce(expected)

      const settings: any = {
        pathToJest: expected,
        rootPath: '',
      }

      expect(pathToJest(settings)).toBe(expected)
      expect(mockNormalize).toBeCalledWith(settings.pathToJest)
    })

    it('defaults to "node_modules/.bin/jest" when Jest is locally installed', () => {
      const expected = 'node_modules/.bin/jest'

      mockJoin.mockImplementation(require.requireActual('path').posix.join)
      mockPlatform.mockReturnValue('linux')
      mockNormalize.mockImplementationOnce(arg => arg)
      mockExistsSync.mockImplementation(path => path === expected)

      expect(pathToJest(defaultSettings)).toBe(expected)
    })

    it('defaults to "jest" when Jest is locally installed', () => {
      const expected = 'jest'

      mockJoin.mockImplementation(require.requireActual('path').posix.join)
      mockPlatform.mockReturnValue('linux')
      mockNormalize.mockImplementationOnce(arg => arg)
      mockExistsSync.mockImplementation(_ => false)

      expect(pathToJest(defaultSettings)).toBe(expected)
    })
  })
})
