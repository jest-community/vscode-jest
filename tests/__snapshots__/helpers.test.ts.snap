// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toErrorString: arbitrary object 1`] = `"{"text":"anything","value":1}"`;

exports[`toErrorString: string 1`] = `"regular error"`;

exports[`toErrorString: undefined 1`] = `""`;
